### PR TITLE
improve error message when `global_asm!` uses `asm!` options

### PR DESCRIPTION
**Pull request by folkertdev**
specifically, what was

    error: expected one of `)`, `att_syntax`, or `raw`, found `preserves_flags`
      --> $DIR/bad-options.rs:45:25
       |
    LL | global_asm!("", options(preserves_flags));
       |                         ^^^^^^^^^^^^^^^ expected one of `)`, `att_syntax`, or `raw`

is now 

    error: the `preserves_flags` option cannot be used with `global_asm!`
      --> $DIR/bad-options.rs:45:25
       |
    LL | global_asm!("", options(preserves_flags));
       |                         ^^^^^^^^^^^^^^^ the `preserves_flags` option is not meaningful for global-scoped inline assembly

mirroring the phrasing of the [reference](https://doc.rust-lang.org/reference/inline-assembly.html#options). 

This is also a bit of a refactor for a future `naked_asm!` macro (for use in `#[naked]` functions). Currently this sort of error can come up when switching from inline to global asm, or when a user just isn't that experienced with assembly. With  `naked_asm!` added to the mix hitting this error is more likely.

### CODE REVIEW

File: compiler/rustc_builtin_macros/src/errors.rs [hunk 1]
```diff
@@ -845,6 +845,17 @@ pub(crate) struct AsmOptAlreadyprovided {
...
 }
 
+#[derive(Diagnostic)]
+#[diag(builtin_macros_global_asm_unsupported_option)]
+pub(crate) struct GlobalAsmUnsupportedOption {
+    #[primary_span]
+    #[label]
+    pub(crate) span: Span,
+    pub(crate) symbol: Symbol,
+    #[suggestion(code = "", applicability = "machine-applicable", style = "tool-only")]
```
Comment (tgross35) [inline compiler/rustc_builtin_macros/src/errors.rs:855]: I think anything machine applicable needs a `//@ run-rustfix` test to make sure the output is correct

File: compiler/rustc_builtin_macros/src/asm.rs [hunk 2]
```diff
@@ -338,25 +351,29 @@ fn parse_options<'a>(
...
+            (sym::att_syntax, ast::InlineAsmOptions::ATT_SYNTAX),
+            (kw::Raw, ast::InlineAsmOptions::RAW),
+        ];
+
+        'blk: {
+            for (symbol, option) in OPTIONS {
+                let expect =
+                    !is_global_asm || ast::InlineAsmOptions::GLOBAL_OPTIONS.contains(option);
+
+                if if expect { p.eat_keyword(symbol) } else { p.eat_keyword_noexpect(symbol) } {
```
Comment (tgross35) [inline compiler/rustc_builtin_macros/src/asm.rs:371]:
```suggestion
                let kw_matched = if !is_global_asm
                    || ast::InlineAsmOptions::GLOBAL_OPTIONS.contains(option) {
                    p.eat_keyword(symbol)
                } else {
                    p.eat_keyword_noexpect(symbol)
                };
                
                if kw_matched { ... }
```

Kind of tough to read the `if if`

**Comment by matthiaskrgr** [comment]
@bors r-
https://github.com/rust-lang/rust/pull/128268#issuecomment-2254137360

**Comment by folkertdev** [comment]
so CI fails on some of the aarch64 tests. I'm on x86, so how am I supposed to update those? for some using `no_core` works, but for `tests/ui/asm/aarch64/parse-error.rs ` it doesn't because that actually used `std` (the `format` macro specifically)

**Comment by Amanieu** [comment]
You can run the UI tests with a cross-compiler by running `./x.py test tests/ui/asm --target aarch64-unknown-linux-gnu --bless`

File: tests/ui/asm/unsupported-option.rs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+//@ run-rustfix
```
Comment (Amanieu) [inline tests/ui/asm/unsupported-option.rs:1]: This probably needs `//@ needs-asm-support` otherwise it will fail on targets without asm support.

**Pull request merged**